### ISSUE TITLE
Bounce behaves strangely with small objects

### ISSUE THREAD

**Issue by mgiraud** [godot-jolt/godot-jolt#374]
Godot 4.02
Windows 10
Intel(R) Core(TM) i7-7700K CPU
Nvidia 2080 Super

Let me give you a little bit of context on how i've found this potential issue. I'm making a pinball using real measurements, meaning that the ball actually has a radius of 1.4cm. The issue that i have is that the ball does not bounce on any part of the pinball even with a material with bounce=1.

After some testing :
* Raising the radius enables the bouncing at some point but i'm not sure it's relevant or more related to the angular or linear speed (Check scene bounce2 on MRP)
* Droping sphere from different heights leads to different behaviours. The weirdest one is the ball bouncing higher and higher (check scene bounce on MRP)

EDIT : Please check the difference with godot physics engine

[bouncing-godot-jolt.zip](https://github.com/godot-jolt/godot-jolt/files/11522202/bouncing-godot-jolt.zip)

**Comment by mihe** [comment]
That does look a bit strange. I'll see if I can reproduce it in Jolt's `Samples` application.

In the meantime, it might be relevant to point out something written under the "Conventions and limits" part of [Jolt's documentation](https://jrouwe.github.io/JoltPhysics/):

> Note that the physics simulation works best if you use SI units (meters, radians, seconds, kg). In order for the simluation to be accurate, dynamic objects should be in the order [0.1, 10] meters long and have speeds in the order of [0, 500] m/s. Static object should be in the order [0.1, 2000] meters long. If you are using different units, consider scaling the objects before passing them on to the physics simulation.

Your objects obviously fall within these conventions, so I wouldn't expect them to behave like this, but you're getting close to having objects that are "too small" for Jolt.

I should really mention this somewhere in the README.

Thanks for all the issues by the way, very much appreciated. Keep them coming. :)

**Comment by mrezai** [comment]
> the ball actually has a radius of 1.4cm

> dynamic objects should be in the order [0.1, 10] meters long

> Your objects obviously fall within these conventions, so I wouldn't expect them to behave like this, but you're getting close to having objects that are "too small" for Jolt.

Ball diameter is 2.8cm so it is smaller than 0.1 meter(10cm) and it isn't within range.
I think it is a good idea that OP makes everything in project 5 times bigger and test it again.

**Comment by mihe** [comment]
> Ball diameter is 2.8cm so it is smaller than 0.1 meter(10cm) and it isn't within range.

Yes, you're right, sorry, I completely skimmed over the numbers and didn't address the dimensions in `bounce2.tscn`. Those are indeed not within the above mentioned range and are likely too small.

The sphere shapes in `bounce.tscn` are however 0.2 m in diameter, and still behave somewhat strangely.

I've managed to reproduce this in Jolt's `Samples` application, so now I'm curious to hear what @jrouwe thinks about this.

Here's the equivalent setup in `Samples`:

```cpp
CreateFloor();

BodyCreationSettings ball1(new SphereShape(0.1f), RVec3(0, 0.15f, 0), Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING);
ball1.mOverrideMassProperties = EOverrideMassProperties::CalculateInertia;
ball1.mMassPropertiesOverride.mMass = 1.0f;
ball1.mRestitution = 1.0f;
ball1.mLinearDamping = 0.1f;
ball1.mAngularDamping = 0.1f;

BodyCreationSettings ball2(new SphereShape(0.1f), RVec3(-0.25f, 0.16f, 0), Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING);
ball2.mOverrideMassProperties = EOverrideMassProperties::CalculateInertia;
ball2.mMassPropertiesOverride.mMass = 1.0f;
ball2.mRestitution = 1.0f;
ball2.mLinearDamping = 0.1f;
ball2.mAngularDamping = 0.1f;

BodyCreationSettings ball3(new SphereShape(0.1f), RVec3(-0.5f, 0.17f, 0), Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING);
ball3.mOverrideMassProperties = EOverrideMassProperties::CalculateInertia;
ball3.mMassPropertiesOverride.mMass = 1.0f;
ball3.mRestitution = 1.0f;
ball3.mLinearDamping = 0.1f;
ball3.mAngularDamping = 0.1f;

mBodyInterface->CreateAndAddBody(ball1, JPH::EActivation::Activate);
mBodyInterface->CreateAndAddBody(ball2, JPH::EActivation::Activate);
mBodyInterface->CreateAndAddBody(ball3, JPH::EActivation::Activate);
```

Which looks like this:

https://github.com/godot-jolt/godot-jolt/assets/4884246/6ac05869-adb2-47e5-9f3d-2ab7e7b324ee

**Comment by jrouwe** [comment]
The behavior is certainly a bit odd. I'd need to step through the simulation to get to a definitive conclusion, but at first glance the objects are indeed quite small and don't work well with the default PhysicsSettings. First of all I set the PhysicsSettings::mSpeculativeContactDistance to 0.005 and mPenetrationSlop to 0.005 too (was 2 cm). But with a body with EMotionQuality::Discrete the body can step into the floor and the next step it is pushed out again (at which point it gains extra velocity). Switching the body to the LinearCast quality and lowering mLinearCastThreshold to 0.2 to ensure that linear casts are actually performed for these small shapes makes the simulation behave better.

**Comment by mihe** [comment]
I only just now noticed that @mgiraud had actually enabled "Continuous CD" (aka `LinearCast`) for the bodies in the repro project attached above, so my port wasn't entirely accurate. Unfortunately I haven't exposed `mLinearCastThreshold` in the project settings, so that's something I need to do.

I can confirm that changing the project settings "Contact Speculative Distance" (aka `mSpeculativeContactDistance`) and "Contact Allowed Penetration" (aka `mPenetrationSlop`) to the values you suggested, along with the above mentioned CCD settings, makes the simulation behave better in the repro project as well.

Although, I just randomly tried setting the linear/angular damping to 0 for the spheres, and that seems to undo some of the effect of the CCD and they once again start bouncing higher and higher.

**Comment by mihe** [comment]
I just released version [0.1.0-rc3](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.1.0-rc3), which includes a new project setting for `mLinearCastThreshold`, as well as a project setting for `mLinearCastMaxPenetration`.

Note that they're both percentages in Godot Jolt, as opposed to factors, so the value of 0.2 for the movement threshold mentioned earlier would instead be 20% in Godot Jolt.

**Comment by jrouwe** [comment]
I've done some further debugging and found a bug in speculative contact resolution, with this fix it looks like:

https://github.com/godot-jolt/godot-jolt/assets/1621693/1213daa5-2461-4fc4-bb4b-2887fa0bab31

The rightmost sphere doesn't bounce because its velocity is lower than PhysicsSettings::mMinVelocityForRestitution. Adding this to the test:

```
PhysicsSettings settings = mPhysicsSystem->GetPhysicsSettings();
settings.mMinVelocityForRestitution = 0.5f;
mPhysicsSystem->SetPhysicsSettings(settings);
```

Makes it look like this:

https://github.com/godot-jolt/godot-jolt/assets/1621693/85264ba5-f7f9-4e9f-9969-71cdd7fae452

You can see it still picks up speed, but I don't think that can be avoided unless you have a restitution < 1 or more linear damping is applied. This is why it happens:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/7317736d-c498-41cf-a82a-22542f75de73)

**Comment by mihe** [comment]
> The rightmost sphere doesn't bounce because its velocity is lower than PhysicsSettings::mMinVelocityForRestitution

Seems like I have another project setting to add. :)

I'll make sure to update Jolt to get the bug fix in.

> You can see it still picks up speed, but I don't think that can be avoided unless you have a restitution < 1 or more linear damping is applied.

Any theories as to why Godot Physics doesn't seem to exhibit this problem, even without CCD? For reference, here's what that looks like:

https://github.com/godot-jolt/godot-jolt/assets/4884246/ef11e7ba-e97f-4827-93e8-41de759afd41

**Comment by jrouwe** [comment]
I honestly have no idea, I tried several theories but didn't find anything.

**Comment by mihe** [comment]
Running with the bugfix for speculative contacts, as well as the new project setting "Bounce Velocity Threshold" (that maps to `mMinVelocityForRestitution`) and the now custom restitution combiner, this is what Godot Jolt looks like if you dial the restitution back to 0.9:

https://github.com/godot-jolt/godot-jolt/assets/4884246/9a085892-95c6-47e3-9453-8b7f34b0f454

Since this issue was mostly about small objects acting strangely, and the restitution 1.0 thing seemingly affecting objects of all sizes, I'll go ahead and close this now.

**Comment by jrouwe** [comment]
The issue that spheres bounce higher and higher should have been fixed by jrouwe/JoltPhysics#1328.

Note that this change needs to be integrated to godot-jolt before it can be tested.

**Comment by mihe** [comment]
The above mentioned change has now been pulled in (#1019).
